File: 518-coin-change-ii/coin-change-ii.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        n = len(coins)
        dp = [[0] * (amount + 1) for _ in range(n)]
        if amount == 0:
            return 1
        for i in range(n):
            for j in range(1, amount+1):
                if j == coins[i]:
                    dp[i][j] = 1
                elif j - coins[i] > 0:
                    dp[i][j] = dp[i][j-coins[i]]
                if i >= 1:
                    dp[i][j] += dp[i-1][j]
        return dp[n-1][amount]

File: 518-coin-change-ii/coin-change-ii.cpp
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        using u128 = unsigned __int128;
        if (amount == 0) return 1;
        if (coins.empty()) return 0;
        
        vector<u128> dp(amount + 1, 0);
        dp[0] = 1;
        
        for (int coin : coins) {
            if (coin <= 0 || coin > amount) continue;
            
            for (int i = coin; i <= amount; i++) {
                dp[i] += dp[i - coin];
            }
        }
        
        return dp[amount];
    }
};